### ISSUE TITLE
Matching array with html to skip re-Rendering 

### ISSUE THREAD

**Issue by SvetoslavStefanov** [knockout/knockout#2127]
I'm having a very slow machine so I have to improvise some optimisations.

I'm having an array with 9 very big objects inside. I can hide the initial render, but after this I have to cache it somehow. So it is possible to cache the DOM object in some (global) variable and the next time when I open the page to use the same DOM object and to match my array to the object ? I tried it, but when I applyBindings and everything goes wrong ... my only solution for now is to re render the dom every single time.

Also are there some optimisations that I can apply to rendering big arrays ?

**Comment by jlaustill** [comment]
@SvetoslavStefanov have you tried using http://knockoutjs.com/documentation/deferred-updates.html

**Comment by SvetoslavStefanov** [comment]
Hi @jlaustill, thank you for your proposal.
I've tried it, but there is a very slight improvement. 
Now rendering takes me about 6-7 seconds, with deferred updates it takes me about 200 ms less ... I'm looking for rendering less than 2 seconds.

Is there a way to match the array to the already rendered DOM ? I want to skip rendering somehow or to make it faster.

I found two things with big impact on rendering:
I'm having a div with many css properties attached through data-bind. It turns out that this is slowing down the rendering.
The second thing is to use "virtual dom" instead of divs for foreach. Like this way -  

But again it's too slow, I made it about 5 seconds with these changes. More ideas ?

**Comment by jlaustill** [comment]
It will be tough to offer more advice without seeing your code.  Can you maybe do up a jsbin or fiddle of one of your objects and how you are updating them?

**Comment by SvetoslavStefanov** [comment]
Hi @jlaustill , 
Sorry for my late answer. Here you can browse the demo that I made today - http://navigation.sstefanov.com/  
The code that I'm using in the project is not the same it's much more complicated. That's why I wrote this sample. The problem is the rendering.
The machine where this code is running is very slow, like 50-70 MB of RAM.

p.s. yts.ag is some random site that I found.

**Comment by jlaustill** [comment]
Are you trying to load the entire dataset all at once like this in your real code?

**Comment by SvetoslavStefanov** [comment]
Yes ?! My idea is to build all the rows, show them with only one item ( _header_ item ) and when the data for each row comes ( request are asyncs ) to be rendered separately. This way I can navigate easily between rows by only showing and hiding them ( with _visible_ in knockout ).
I've tried a different approach - to render just 3 rows, but this way when I have to go to the 4th one for example, I have to hide the first one and show the next one. But this means I have to _shift_ the first row and to _push_ the last one, which is append / prepend action in JS and it's visibly slower.

Do you have any different proposal than these two ?

**Comment by cervengoc** [comment]
Please have a look at https://github.com/brianmhunt/knockout-fast-foreach, it still creates the DOM nodes, but includes a lot of optimizations compared to built-in `foreach`. Take care though, the two is not completely the same when it comes to features and behavior. However, in simple use-cases you just have to change the binding key to 'fastForEach' and everything works the same.

**Comment by SvetoslavStefanov** [comment]
HI @cervengoc,
I have looked at fastForEach, but didn't managed to make it run on that machine. The _machine_ actually is a set top box with unknown version of webkit. So I can't see when something trows errors, I just see that nothing happens and I couldn't debbug it.

I will give it a try again after the holidays, but I'm not very optimistic about it.

**Comment by SvetoslavStefanov** [comment]
I've tried fastForEach and it works ! Now the loading time is about 4 seconds. Thanks cervengoc !

Any ideas are welcomed.

**Comment by SvetoslavStefanov** [comment]
I've made an improvement today ! Rather than changing the _visible_ property, I used _if_ . This way if the row is not visible, it wouldn't be rendered ( also it's items ). This way the loading became about 2 seconds, but now scrolling between rows looks sluggish ...

I'll keep updating this thread with things that I found.
